src/engine/blocks.js

```diff
@@ -25,6 +25,20 @@ class Blocks {
          * @type {Array.<String>}
          */
         this._scripts = [];
+
+        /**
+         * Cache procedure Param Names by block id
+         * @type {object.<string, ?Array.<string>>}
+         * @private
+         */
+        this._procedureParamNames = {};
```

Comment:
This PR has two caching patterns. Storing cached values on blocks themselves and storing cached values on the Blocks instance.

I think you can reduce this to one pattern and simplify cache invalidation.

Stored cached values in category members of a cache member on Blocks.

Like maybe:

```js
this._cached = {
  inputs: {},
  procedureParamNames: {},
  procedureDefinitions: {},
};
```

Then have a method `_resetCache` that resets `_cached` to the same object expression.

You can call `_resetCache` at the end of `addBlock`, `changeBlock`, `moveBlock`, and `deleteBlock` to reset the caches.